src/lib/phdi-building-blocks/phdi_building_blocks/schemas.py

```diff
@@ -0,0 +1,251 @@
+    :param schema: A schema specifying the desired values by FHIR resource type.
+    :param credential_manager: A credential manager for
+    a FHIR server.
+    """
+
+    output_path.mkdir(parents=True, exist_ok=True)
+    output_file_name = output_path / f"{resource_type}.{output_format}"
+
+    query = f"/{resource_type}"
+    response = query_fhir_server(credential_manager, query)
+
+    additional_page = True
+    writer = None
+    while additional_page:
+        if response.status_code != 200:
+            log_fhir_server_error(response.status_code)
+            break
+
+        # Load queried data.
+        query_result = json.loads(response.content)
+        raw_schema_data = []
+
+        # Extract values specified by schema from each resource.
+        for resource in query_result["entry"]:
+            values_from_resource = apply_schema_to_resource(
+                resource["resource"], schema
+            )
+            if values_from_resource != {}:
+                raw_schema_data.append(values_from_resource)
+
+        # Write data to parquet
+        writer = write_schema_table(
+            raw_schema_data, output_file_name, output_format, writer
+        )
+
+        # Check for an additional page of query results.
+        for link in query_result.get("link"):
+            if link.get("relation") == "next":
+                next_page_url = link.get("url")
+                response = query_fhir_server(
+                    credential_manager, specific_url=next_page_url
+                )
+                break
+            else:
+                response = None
+
+        if response is None:
+            additional_page = False
+    if writer is not None:
+        writer.close()
+
+
+def generate_schema(
+    fhir_url: str,
+    schema_path: pathlib.PosixPath,
+    output_path: pathlib.PosixPath,
+    output_format: Literal["parquet"],
+):
+    """
+    Given the url for a FHIR server, the location of a schema file, and and output
+    directory generate the specified schema and store the tables in the desired
+    location.
+
+    :param fhir_url: URL to a FHIR server.
+    :schema_path: A path to the location of a YAML schema config file.
+    :output_path: A path to the directory where tables of the schema should be written.
+    :output_format: Specifies the file format of the tables to be generated.
+    """
+
+    schema = load_schema(schema_path)
+    schema_name = list(schema.keys())[0]
+    schema = schema[schema_name]
+    output_path = pathlib.Path(schema_name)
+
+    credential_manager = AzureFhirserverCredentialManager(fhir_url)
+
+    for resource_type in schema.keys():
+        make_resource_type_table(
+            resource_type, schema, output_path, output_format, credential_manager
+        )
+
+
+def write_schema_table(
+    data: List[dict],
+    output_file_name: pathlib.PosixPath,
+    file_format: Literal["parquet"],
+    writer: pq.ParquetWriter = None,
```

Comment:
This function should not take in a `writer` object. It should open a connection to the file itself, long enough to write the data that it needs to, and then close the connection. The way this is currently implemented keeps the connection open longer than necessary.

Reply:
UPDATE: Just saw the comment below about not being able to append to a Parquet file. I think this is fine for now, but please add a `#TODO` comment stating that we need to figure out how to overcome this. A couple solutions I've found online require reading the existing file in, appending the new data, and writing back out. That's a lot of IO, and we'd need to consider the costs of that in the cloud, but something to consider.

I think the ideal solution would be to simply batch all of the data together and make one write operation. This will require some analysis around data sizes, but could be a viable solution.

Reply:
We could also write many files containing chucks of a single table. Then read them each in turn and write to a new file. I think this would require less IO than the initial proposal, but not require ever holding all the data in memory.